src/routers/transactions.py

```diff
@@ -60,10 +67,37 @@ def read_transactions(
         )
 
     # date range filter
-    # TODO
+    if query_params.start_date is not None:
+        query = query.where(Transaction.trans_date >= query_params.start_date)
+    if query_params.end_date is not None:
+        query = query.where(Transaction.trans_date <= query_params.end_date)
+
+    # get total record count
+    count_query = select(func.count(1).label("cnt")).select_from(query.subquery())
+    total_row_count = session.exec(count_query).one()
+
+    # calculate total page count
+    total_page_count = (
+        total_row_count + pagination_input.size - 1
+    ) // pagination_input.size
+
+    # determine actual page to give; give last page if requested page is out of bounds
```

Comment:
but yeah we can break this up a bit more, usually inline comments hint at the need to do:

Extract Method Refactoring Example in Python
https://www.youtube.com/watch?v=N9sqd1-0_qk